File: packages/protocol-parser/src/abiTypesToSchema.ts
import { DynamicAbiType, SchemaAbiType, StaticAbiType, isDynamicAbiType } from "@latticexyz/schema-type";
import { Schema } from "./common";

export function abiTypesToSchema(abiTypes: SchemaAbiType[]): Schema {
  const staticFields: StaticAbiType[] = [];
  const dynamicFields: DynamicAbiType[] = [];
  for (const abiType of abiTypes) {
    if (isDynamicAbiType(abiType)) dynamicFields.push(abiType);
    else staticFields.push(abiType);
  }
  return { staticFields, dynamicFields };
}

File: packages/services/pkg/protocol-parser/abiTypesToSchema.go
package protocolparser

import schematype "latticexyz/mud/packages/services/pkg/schema-type"

func AbiTypesToSchema(abiTypes []schematype.SchemaType) Schema {
	staticFields := []schematype.SchemaType{}
	dynamicFields := []schematype.SchemaType{}

	for _, abiType := range abiTypes {
		if abiType.IsDynamic() {
			dynamicFields = append(dynamicFields, abiType)
		} else {
			staticFields = append(staticFields, abiType)
		}
	}

	return Schema{
		StaticFields:  staticFields,
		DynamicFields: dynamicFields,
	}
}
